crates/js-component-bindgen/src/transpile_bindgen.rs

```diff
@@ -325,9 +336,32 @@ impl<'a> JsBindgen<'a> {
                 "\
                     {}
                     {}
-                    {maybe_init_export}const $init = (async() => {{
-                        {}\
-                        {}\
+                    {maybe_init_export}const $init = (() => {{
+                        let gen = (function* init () {{
+                            {}\
+                            {}\
+                        }})();
+                        let promise, resolve, reject;
+                        function runNext (value) {{
+                            try {{
+                                let done;
+                                do {{
+                                    ({{ value, done }} = gen.next(value));
+                                }} while (!(value instanceof Promise) && !done);
+                                if (done) {{
+                                    if (resolve) resolve(value);
+                                    else return value;
+                                }}
+                                if (!promise) promise = new Promise((_resolve, _reject) => (resolve = _resolve, reject = _reject));
+                                value.then(nextVal => done ? resolve() : runNext(nextVal), reject);
```

Comment:
Since the done path is handled above, this should probably just be:

```suggestion
                                value.then(runNext, reject);
```